WordVectors/python/analysis_notebook.ipynb

```diff
@@ -0,0 +1,892 @@
+    "random.seed(0)\n",
+    "\n",
+    "# grab all of the indices for our model vocabulary\n",
+    "all_indices = list(range(len(labels)))\n",
+    "\n",
+    "# take a random sample of 25 indices\n",
+    "selected_indices = random.sample(all_indices, 25) # you can change this\n",
+    "\n",
+    "# iterate through our random sample\n",
+    "    for i in selected_indices:\n",
+    "        \n",
+    "        # label the corresponding random points\n",
+    "        ax.text(x_axis[i], y_axis[i], labels[i], style ='italic',\n",
+    "        fontsize = 10, color =\"blue\")\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The entire `pca()` function with a random sample of 25 items labeled is below:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "labels = list(model.wv.key_to_index)\n",
+    "def pca(model):\n",
+    "    X = model.wv[model.wv.key_to_index]  # get all the vectors\n",
+    "    pca = PCA(n_components=2)\n",
+    "    result = pca.fit_transform(X)\n",
+    "    # create a scatter plot of the projection\n",
+    "    x_axis = result[:,0]\n",
+    "    y_axis = result[:,1]\n",
+    "    fig = plt.figure(figsize=(8, 6))\n",
+    "    ax = fig.add_subplot(111)\n",
+    "    ax.scatter(x_axis, y_axis)\n",
+    "\n",
+    "    import random\n",
+    "    random.seed(0)\n",
+    "    all_indices = list(range(len(labels)))\n",
+    "    selected_indices = random.sample(all_indices, 25) # you can change this\n",
+    "    for i in selected_indices:\n",
+    "        ax.text(x_axis[i], y_axis[i], labels[i], style ='italic',\n",
+    "        fontsize = 10, color =\"blue\")\n",
+    "\n",
+    "    \n",
+    "    plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "PCA plots can also be represented three-dimensionally which can often provide a different view of your model vocabulary. With the PCA plot represented three-dimensionally, you can get a better sense of the shape of the plot and where words are located in that space. The code is nearly the same as the two-dimensional plot above. The key difference between the two, is that instead of plotting an X and Y axis, we are going to plot a X, Y, and Z axis. The Z axis will represent the third dimension that we are adding to our plot.\n",
+    "\n",
+    "Everything is nearly the same but you'll notice that in addition to defining the `x_axis` and `y_axis` variables, we are also going to define a variable `z_axis` which will pull vectors from the third column in our `result` variable."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# define x axis\n",
+    "x_axis = result[:,0]\n",
+    "\n",
+    "# define y axis\n",
+    "y_axis = result[:,1]\n",
+    "\n",
+    "# define z axis\n",
+    "z_axis = result[:,2]"
```

Comment:
To be honest, about here I stopped trying to run _any_ code blocks and just skimmed the remainder of the notebook. It’s exhausting and frustrating to try to figure out from context whether this is a code block I can run or not. This cell looks simple enough, but others did too and I got only errors. I feel a bit like a mouse who got shocked too many times and is now unwilling to push the button that may or may not release a treat?

I’m not sharing this to chide, but because when I reflected on how I was feeling, I realized that it’s valuable feedback. There’s a real expectation with interactive code books like this that everything will work out of the box. Errors don’t feel like challenges to the novice, they feel like judgments of character. It’s exhausting to fight through to try to find a useful and applicable tidbit. I came to this with a little more coding experience and a little more commitment to getting through the notebook than our target audience, but even so, I did get exhausted. :/

Reply:
I'm sorry that this happened :( I was thinking from the perspective of putting the explanation of the line with the cooresponding line of code, but maybe that format doesn't work well with these function definitions because you need all of the code to run together for it to work. Do you think I should keep the format and just signpost more, or should I put all of the explanation together and then the code?

Reply:
Quick summary of what we discussed on Slack: Stepping through the code to explain it is great, and more so if running the code can demonstrate or reinforce the paragraphs of descriptions. To that end, the functions can be assembled in whole after their contents are explained piece-by-piece.